walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,108 @@
+        Collections.sort(eventsByEndTime, Event.ORDER_BY_END_DESCENDING);
+        Collections.sort(eventsByStartTime, Event.ORDER_BY_START_ASCENDING);
+
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        /**
+         * To find valid meeting times, we use a temporary variable prevEndTime
+         * that is initially set to the end of the day. Since we sort events by 
+         * their end time in descending order, we effectively travel backwards.
+         * Starting with the latest ending event, check if the event's attendees
+         * overlap with the meeting's required attendees. If they do, then we 
+         * cap off our current time range, which spans [event end time, prevEndTime),
+         * and add it to the valid meeting times. Set prevEndTime to the start
+         * of the conflicting event and continue on to the next event. If there 
+         * are no intersecting attendees, simply move on to next event.
+         */
+        int prevEndTime = TimeRange.END_OF_DAY + 1;
+        for (Event event : eventsByEndTime) {
+            /* Find attendees required to be at meeting that are also attending event. */
+            Collection<String> eventAttendees = event.getAttendees();
+            Collection<String> conflictAttendees = new HashSet<>(attendees);
+            conflictAttendees.retainAll(event.getAttendees());
+
+            if (conflictAttendees.size() > 0) {
+                TimeRange freeTime = TimeRange.fromStartEnd(event.getWhen().end(), prevEndTime, false);
+                if (duration <= freeTime.duration()) {
+                    meetingTimes.add(freeTime);
+                } 
+                prevEndTime = (event.getWhen().start() < prevEndTime) ? event.getWhen().start() : prevEndTime;
+            }
+        }
+
+        /* Check the time range from the start of the day to the first event's start time. */
+        if (prevEndTime > 0) {
+            TimeRange firstFreeTime = TimeRange.fromStartEnd(0, prevEndTime, false);
+            meetingTimes.add(firstFreeTime);
+        }
+
+        List<TimeRange> meetingTimesList = new ArrayList<TimeRange>(meetingTimes);
+        Collections.reverse(meetingTimesList);
+        return meetingTimesList;
+    }
+
+    /**
+     * The query function returns a set of non-overlapping TimeRanges where a 
+     * potential meeting, specified as a MeetingRequest argument, can take place 
+     * among a day of Events. A MeetingRequest has a specified duration and set 
+     * of attendees, which can be broken down into a set of mandatory attendees, 
+     * which must attend, and set of optional attendees, that are preferred to 
+     * attend but might not necessarily have to. The algorithm runs in O(n^2) 
+     * time since we have to sort the events first, requiring O(nlogn) time, and 
+     * then iterate through each event and find conflicting attendees, requiring 
+     * O(n^2) time.
```

Comment:
I agree that the sorting will be O(_n_*log(_n_)) where _n_ is the number of events but it's not so obvious why detecting conflicting attendees is O(_n_^2). The number of possible attendees is different from the number of events so I think we should use a different variable (_m_?) for them. Also, I suspect the complexity also depends on how efficient the implementation for `retainAll` is.

Reply:
I remember checking and `retainAll` has time complexity O(_n_)